plone/schemaeditor/browser/schema/add_field.py

```diff
@@ -63,35 +64,45 @@ def create(self, data):
 
         return field_obj
 
-    def add(self, field):
-        context = self.context
-        schema = IEditableSchema(context.schema)
-
-        # move it after the last field that is not in a fieldset
-        # or at top if there is no field yet in "default" fieldset
-        ordered_fields = [name for (name, f) in sortedFields(context.schema)]
-        default_fields = non_fieldset_fields(context.schema)
-        if len(default_fields) > 0:
-            position = ordered_fields.index(default_fields[-1]) + 1
-        else:
-            position = 0
+    def add(self, new_field):
+        schema = self.context.schema
+        fieldset_id = int(self.request.form.get('fieldset_id', 0))
+        position = new_field_position(schema, fieldset_id)
 
+        editable_schema = IEditableSchema(schema)
         try:
-            schema.addField(field)
+            editable_schema.addField(new_field)
         except ValueError:
             raise WidgetActionExecutionError(
                 '__name__',
                 Invalid(
                     u'Please select a field name that is not already used.'
                 )
             )
+        if fieldset_id:
+            fieldset = get_fieldset_from_index(schema, fieldset_id)
+            editable_schema.changeFieldFieldset(new_field.__name__, fieldset)
+        editable_schema.moveField(new_field.__name__, position)
 
-        schema.moveField(field.__name__, position)
-        notify(ObjectAddedEvent(field, context.schema))
-        notify(FieldAddedEvent(context, field))
+        notify(ObjectAddedEvent(new_field, schema))
+        notify(FieldAddedEvent(self.context, new_field))
         IStatusMessage(self.request).addStatusMessage(
             _(u'Field added successfully.'), type='info')
 
+    def updateWidgets(self):
+        super(FieldAddForm, self).updateWidgets()
+        fieldset_id = int(self.request.form.get('fieldset_id', 0))
+        if fieldset_id:
+            # add fieldset_id from GET parameter as hidden field, so that
+            # ``add`` method at the end of the form lifecycle can read it.
+            fieldset_id_widget = TextWidget(self.request)
+            fieldset_id_widget.name = 'fieldset_id'
+            fieldset_id_widget.value = fieldset_id
+            fieldset_id_widget.mode = HIDDEN_MODE
+            # Uhm. z3c.form widgets doesn't have an API for extending a
+            # schema-generated form. Using internal ``_data_values``...
+            self.widgets._data_values.append(fieldset_id_widget)
```

Comment:
I'm not too happy having to use internal data structures just to inject a non-schema related field/widget into an z3c.form - but I didn't find another way.
@davisagli & @jensens maybe you have a better idea how to add a hidden field, which should not part of the schema to a z3c.form?